### ISSUE TITLE
REPL: reloading ES6 modules

### ISSUE THREAD

**Issue by borkdude** [borkdude/cherry#25]
What approaches would be feasible, while also allowing for tree-shaking by ES6 tooling?

**Comment by borkdude** [comment]
One possible approach: in `npx cherry --dev` mode we could emit for every var:

```
var foo = ...
var set_foo = ...
```

Then when evaluating `(defn foo [])` in the REPL, we construct `foo` in a temporary module and then call `original_module.set_foo(foo)`.

**Comment by borkdude** [comment]
Another approach might be: every .cljs namespaces defines a global object (only in development) to facilitate the REPL.

**Comment by lilactown** [comment]
I've thought about this before; I haven't tried it yet, but I think you don't have to map every public var to an `export`. Instead, you could export a mutable object at dev time. pseudo code below

```javascript
/* @cherry-namespace: myapp_main */
import {exports as cljs_core} from "@cherry/core";
import {exports as myapp_feature} from "./feature.js";

function foo () {
  return cljs_core.inc(myapp_feature.magic_number);
}

global.myapp_main = { foo };

export const exports = global.myapp_main;
```

So essentially, during development you would import and export any CLJS source files using the above method, and reloads would update the `global.myapp_main` object. At release time, you could remove the global and change it to exporting each public var, which would probably help with tree shaking (again, haven't tested this).

The problem with this approach would be if you were mixing CLJS and other local ESM code at dev time, because the import would be different depending on whether you were in dev or release mode. I wonder if this could be solved by some sort of plugin for webpack, etc?

**Comment by jo-sm** [comment]
What may work is mimicking how JS tools handle hot module reloading, which is to use `import.meta.hot` and `import.meta.accept`, which  you could then somehow hook into when loading a namespace in a CLJS REPL and rewrite the code. [Here are the docs on how Vite handles it](https://vitejs.dev/guide/api-hmr.html). 

If this approach works, then it mean that the way you deal with HMR and REPL reloading would be basically identical, which would be great and make it simple to work with existing HMR setups in JS-land.

**Comment by bhauman** [comment]
I'd like to chime in that it would be nice to preserve the Clojure semantics around reloading in general.  This behavior being one of the killer features of Clojure.  I.E. reloading a file redefines the vars globally and any side effecting code gets executed.  I do think this is probably best accomplished via compiling in a dev format versus a prod format.

**Comment by borkdude** [comment]
Agreed.

Btw, I know it's not a REPL, but made a little browser playground here:

https://borkdude.github.io/cherry/index.html

**Comment by bhauman** [comment]
Here's an interesting approach to reloading:  https://itnext.io/hot-reloading-native-es2015-modules-dc54cd8cca01

The interesting thing is calling the "import" function to reload a file using a different URL (using a query param) to force re-evaluation of the file.  Thus the top level stuff gets evaled and if you place the module functions on a global context variable `window.cherry.root.example.foo = function(a,b) {...}` they will be reassigned on loading.

**Comment by borkdude** [comment]
> The interesting thing is calling the "import" function to reload a file using a different URL (using a query param) to force re-evaluation of the file. 

It's also used in the playground example, so executing the same snippet twice will evaluate twice. The snippet is first compiled and then encoded as a data url and then imported by the browser. :)

**Comment by bhauman** [comment]
Very cool! I hadn't looked into this before!

**Comment by borkdude** [comment]
I think @lilactown's approach makes sense too:

https://github.com/borkdude/cherry/issues/25#issuecomment-1203238677

**Comment by bhauman** [comment]
Yeah that looks good! I guess it goes without saying that all calls have to be indirect through the mutable export.

An alternative would be to encapsulate the mutable exports in its own module and not make it global. The module could have a convenience method to attach it to the global object for debugging

**Comment by borkdude** [comment]
> Actually I might place the module on a common root object and then in turn place that on the global object.

With module, do you mean the mutable object with vars, or an ES6 module?

Some pseudo-code would help me understand this better.

> An alternative would be to encapsulate the mutable exports in its own module

Again, some pseudo-code would help, I think I'm not following exactly.

**Comment by bhauman** [comment]
Yeah I should have provided examples to start with.

Here's the above example with a root object as in intermediary to the global object. This helps reduce the likelihood global name collision and makes it simpler matter to clean things up for a full reset without hitting reload.

```javascript
/* @cherry-namespace: myapp_main */
import {exports as cljs_core} from "@cherry/core";
import {exports as myapp_feature} from "./feature.js";

function foo () {
  return cljs_core.inc(myapp_feature.magic_number);
}
/* EDIT */
/* ROOT object to avoid name collisions and to aid with clean up of live env if necessary. */
global.cherry_env_root_object.myapp_main = { foo };

export const exports = global.cherry_env_root_object.myapp_main;
```

Here's a further example where we encapsulate a mutable root object. Encapsulating the root object further reduces any chance of interference from whatever else is executing in the runtime env.

```javascript
/* cherry_env_root */

const context = {cherry_env_root: {} };

export default  context.cherry_env_root;
```

```javascript
/* @cherry-namespace: myapp_main */
import {exports as cljs_core} from "@cherry/core";
import {exports as myapp_feature} from "./feature.js";

/* EDIT import encapsulated root from module */
import {exports as cherry_root} from "./cherry_env_root.js";



function foo () {
  return cljs_core.inc(myapp_feature.magic_number);
}
/* EDIT */
/* ROOT object to avoid name collisions and to aid with clean up of live env if necessary. */
global.cherry_root.myapp_main = { foo };

export const exports = global.cherry_root.myapp_main;
```

**Comment by borkdude** [comment]
Much clearer, thank you!

**Comment by borkdude** [comment]
For completeness, just spitting out more ideas:

One could emit vars as var foo_fn = {val: function() { } } so vars are always mutable, also from the outside (ES6-wise). calls to vars would be compiled as foo_fn.val(1, 2, 3) (a var deref, like in Clojure JVM). This comes with a little bit of performance overhead, but maybe not that much. This will solve the REPL and dynamic var problem I think.

The annoying bit is when calling these suckers from JS: you also have to do the .val indirection, which sucks. Darn.

Or, what if you had:
```
> var foo_fn = function(...args) { return foo_fn.val(...args) }
> foo_fn.val = function(...args) { return [...args] }
> foo_fn(1, 2, 3)
[ 1, 2, 3 ]

> foo_fn.val = function(...args) { return args[0] }
[Function (anonymous)]
> foo_fn(1, 2, 3)
1
```

This only works for functions though.

**Comment by borkdude** [comment]
Some progress from squint: https://twitter.com/borkdude/status/1577255756919115777
This will be ported to cherry as well (I'm planning to unify the compiler code of both projects)

**Comment by borkdude** [comment]
More news:

https://github.com/squint-cljs/squint#repl
